Rakefile

```diff
@@ -48,15 +51,64 @@ RuboCop::RakeTask.new('lint') do |t|
   t.fail_on_error = false
 end
 
-Gem::PackageTask.new(gem) do |p|
+Gem::PackageTask.new(RUBY_MAGIC_GEM_SPEC) do |p|
   p.need_zip = false
   p.need_tar = false
 end
 
-Rake::ExtensionTask.new('magic', gem) do |e|
+CROSS_RUBY_PLATFORMS = ["x86_64-linux", "x86-linux"]
+CROSS_RUBY_VERSIONS = ["2.5.0", "2.6.0", "2.7.0", "3.0.0"].join(":")
+
+require "rake_compiler_dock"
+
+ENV["RUBY_CC_VERSION"] = CROSS_RUBY_VERSIONS
+
+Rake::ExtensionTask.new('magic', RUBY_MAGIC_GEM_SPEC) do |e|
   e.source_pattern = '*.{c,h}'
   e.ext_dir = 'ext/magic'
   e.lib_dir = 'lib/magic'
+
+  e.cross_compile = true
+  e.cross_config_options << "--enable-cross-build"
+  e.cross_platform = CROSS_RUBY_PLATFORMS
+  e.cross_compiling do |spec|
+    spec.files.reject! { |path| File.fnmatch?('ports/*', path) }
+    spec.files.reject! { |path| File.fnmatch?('patches/*', path) }
+    spec.dependencies.reject! { |dep| dep.name=='mini_portile2' }
+
+    db_glob = "ext/magic/share/*.mgc"
+    raise "magic files are not present at #{db_glob}" if Dir[db_glob].empty?
+    Dir[db_glob].each { |db| spec.files << db }
+  end
+end
+
+namespace "gem" do
+  def gem_builder(platform)
+    # use Task#invoke because the pkg/*gem task is defined at runtime
+    Rake::Task["native:#{platform}"].invoke
+    Rake::Task["pkg/#{RUBY_MAGIC_GEM_SPEC.full_name}-#{Gem::Platform.new(platform).to_s}.gem"].invoke
```

Comment:
On macOS, I'm seeing:

```
/usr/local/rake-compiler/ruby/x86_64-redhat-linux/ruby-3.0.0/include/ruby-3.0.0/ruby/internal/memory.h:254:57: warning: conversion to ‘size_t’ {aka ‘long unsigned int’} from ‘long int’ may change the sign of the result [-Wsign-conversion]
  254 |         store, rbimpl_size_mul_or_raise(count, elsize), count);
      |                                                         ^~~~~
linking shared-object magic/magic.so
WARNING:  See https://guides.rubygems.org/specification-reference/ for help
rake aborted!
Gem::InvalidSpecificationException: ["ext/magic/share/magic.mgc"] are not files
/Users/stanhu/github/ruby-magic/rakefile:89:in `gem_builder'
/Users/stanhu/github/ruby-magic/rakefile:105:in `block (4 levels) in <top (required)>'
/usr/local/rvm/gems/ruby-2.5.8/gems/rake-13.0.3/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.8/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.8/bin/ruby_executable_hooks:24:in `<main>'
Tasks: TOP => pkg/ruby-magic-0.4.0-x86_64-linux.gem
(See full trace by running task with --trace)
rake aborted!
Command failed with status (1): [docker run -v /Users/stanhu/github/ruby-magic:/Users/stanhu/github/ruby-magic -e UID\=1000 -e GID\=1000 -e USER\=stanhu -e GROUP\=_staff -e GEM_PRIVATE_KEY_PASSPHRASE -e ftp_proxy -e http_proxy -e https_proxy -e RCD_HOST_RUBY_PLATFORM\=x86_64-darwin19 -e RCD_HOST_RUBY_VERSION\=2.7.2 -e RCD_IMAGE\=larskanis/rake-compiler-dock-mri-x86_64-linux:1.1.0 -w /Users/stanhu/github/ruby-magic --rm -i -t larskanis/rake-compiler-dock-mri-x86_64-linux:1.1.0 runas sigfw bash -c gem\ install\ bundler\ --no-document\ \&\&'
'bundle\ \&\&'
'bundle\ exec\ rake\ gem:x86_64-linux:builder\ MAKE\=\'nice\ make\ -j\`nproc\`\''
']
/Users/stanhu/github/ruby-magic/Rakefile:95:in `block (3 levels) in <top (required)>'
/Users/stanhu/.rbenv/versions/2.7.2/bin/bundle:23:in `load'
/Users/stanhu/.rbenv/versions/2.7.2/bin/bundle:23:in `<main>'
Tasks: TOP => gem:x86_64-linux
(See full trace by running task with --trace)
```

Reply:
You may want to make sure you've blown away your `tmp` and `pkg` directories (my preference would be to have `tmp` in `CLEAN` rather than `CLOBBER` for this reason).

Reply:
Ok, I ran a `rake clobber` and bumped my `VERSION` to 0.4.1 and got the `.gem` to build. After I installed it, I tried to use it:

```ruby
irb(main):001:0> require 'magic'
Traceback (most recent call last):
       10: from /Users/stanhu/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        9: from /Users/stanhu/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        8: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        7: from (irb):1
        6: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:156:in `require'
        5: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:168:in `rescue in require'
        4: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:168:in `require'
        3: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic.rb:3:in `<top (required)>'
        2: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic.rb:8:in `rescue in <top (required)>'
        1: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic.rb:8:in `require_relative'
LoadError (cannot load such file -- /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic/magic)
```

Something isn't working here:

```ruby
begin
  ::RUBY_VERSION =~ /(\d+\.\d+)/
  require_relative "magic/#{Regexp.last_match(1)}/magic"
rescue LoadError => e
  require_relative 'magic/magic'
end
```

If I debug the exception in that `require` block, I see:

```
cannot load such file -- /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic/2.7/magic
```

I noticed if I renamed the `require_relative` calls to `require` above (as Nokogiri seems to do), it works?

Reply:
The reference to `lib/magic/2.7/magic` is to the file `lib/magic/2.7/magic.so` (not to a directory) which it looks like does exist on your system. I'm sorry, I don't fully understand what's going on on your system, and I'm having trouble reproducing it.

Reply:
Wait, are you trying to use the precompiled linux gem on MacOS?

![image](https://user-images.githubusercontent.com/8207/114392165-63e73b80-9b66-11eb-829f-6498f5f394e4.png)

Reply:
🤦 Silly me! I was just testing you.

Re-tested this on a Linux system, and it worked fine. Thanks!